tests/unit/lib/upload_manager_test.py

```diff
@@ -119,20 +141,69 @@ def test_url_tar_gz_should_not_simplify_archives(self):
+        # This test hits the real GitHub repository. If the contents of README.md at https://github.com/codalab/test
+        # change, then update this test.
+        self.check_file_equals_string('testfile.md', '# test\nUsed for testing\n')
 
     def write_string_to_file(self, string, file_path):
         with open(file_path, 'w') as f:
             f.write(string)
 
-    def write_bytes_to_file(self, bytes_, file_path):
-        with open(file_path, 'wb') as f:
-            f.write(bytes_)
 
-    def check_file_contains_string(self, file_path, string):
+class UploadManagerDiskStorageTest(UploadManagerTestBase, unittest.TestCase):
+    """Tests for UploadManager that upload files to disk storage."""
+
+    @property
+    def use_azure_blob_beta(self):
+        return False
+
+    def check_file_equals_string(self, file_subpath: str, expected_contents: str):
+        file_path = (
+            os.path.join(self.bundle_location, file_subpath)
+            if file_subpath
+            else self.bundle_location
+        )
         self.assertTrue(os.path.isfile(file_path))
         with open(file_path, 'r') as f:
-            self.assertEqual(f.read(), string)
+            self.assertEqual(f.read(), expected_contents)
+
+    def listdir(self):
+        return os.listdir(self.bundle_location)
+
+
+class UploadManagerBlobStorageTest(UploadManagerTestBase, unittest.TestCase):
+    """Tests for UploadManager that upload files to Blob Storage."""
+
+    @property
+    def use_azure_blob_beta(self):
+        return True
+
+    def check_file_equals_string(self, file_subpath: str, expected_contents: str):
+        with FileSystems.open(
+            self.bundle_location, compression_type=CompressionTypes.UNCOMPRESSED
+        ) as f:
+            if not file_subpath:
+                # Should be a .gz file
+                self.assertTrue(self.bundle_location.endswith("contents.gz"))
+                self.assertEqual(gzip.decompress(f.read()).decode(), expected_contents)
+            else:
+                # Should be a .tar.gz file
+                self.assertTrue(self.bundle_location.endswith("contents.tar.gz"))
+                with tarfile.open(fileobj=f, mode='r:gz') as tf:
+                    # Prepend "./" to the file subpath so that it corresponds with a file in the archive.
+                    self.assertEqual(
+                        cast(IO[bytes], tf.extractfile("./" + file_subpath)).read().decode(),
+                        expected_contents,
+                    )
+
+    def listdir(self):
+        with FileSystems.open(
+            self.bundle_location, compression_type=CompressionTypes.UNCOMPRESSED
+        ) as f, tarfile.open(fileobj=f, mode='r:gz') as tf:
+            # Remove "." entry and "./" prefixes to make the file listing similar to that
+            # in an ordinary directory on disk.
+            return [i.replace("./", "") for i in tf.getnames() if i != "."]
```

Comment:
why do we have the `if` here? that's kind of implicit, right?

Reply:
We want to not return the `.` entry, so that the file listing is similar to that in an ordinary directory that one gets by running `os.listdir`. See https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions for more about the syntax.

Reply:
oh, I see. List comprehensions trip me up.